src/crimson/osd/shard_services.cc

```diff
@@ -0,0 +1,230 @@
+					    std::move(notifies));
+	logger().debug("dispatch_context_messages sending notify to {}", peer);
+	return send_to_osd(peer, m, osdmap->get_epoch());
+      }),
+    seastar::parallel_for_each(std::move(ctx.query_map),
+      [this](auto& osd_queries) {
+	auto& [peer, queries] = osd_queries;
+	auto m = make_message<MOSDPGQuery>(osdmap->get_epoch(),
+					   std::move(queries));
+	logger().debug("dispatch_context_messages sending query to {}", peer);
+	return send_to_osd(peer, m, osdmap->get_epoch());
+      }),
+    seastar::parallel_for_each(std::move(ctx.info_map),
+      [this](auto& osd_infos) {
+	auto& [peer, infos] = osd_infos;
+	auto m = make_message<MOSDPGInfo>(osdmap->get_epoch(),
+					  std::move(infos));
+	logger().debug("dispatch_context_messages sending info to {}", peer);
+	return send_to_osd(peer, m, osdmap->get_epoch());
+      }));
+  ctx.notify_list.clear();
+  ctx.query_map.clear();
+  ctx.info_map.clear();
+  return ret;
+}
+
+seastar::future<> ShardServices::dispatch_context(
+  ceph::os::CollectionRef col,
+  PeeringCtx &&ctx)
+{
+  ceph_assert(col || ctx.transaction.empty());
+  return seastar::do_with(
+    PeeringCtx{ctx},
+    [this, col](auto& todo) {
+      return seastar::when_all_succeed(
+	dispatch_context_messages(todo),
+	col ? dispatch_context_transaction(col, todo) : seastar::now());
+    });
+}
+
+void ShardServices::queue_want_pg_temp(pg_t pgid,
+				    const vector<int>& want,
+				    bool forced)
+{
+  auto p = pg_temp_pending.find(pgid);
+  if (p == pg_temp_pending.end() ||
+      p->second.acting != want ||
+      forced) {
+    pg_temp_wanted[pgid] = {want, forced};
+  }
+}
+
+void ShardServices::remove_want_pg_temp(pg_t pgid)
+{
+  pg_temp_wanted.erase(pgid);
+  pg_temp_pending.erase(pgid);
+}
+
+void ShardServices::_sent_pg_temp()
+{
+#ifdef HAVE_STDLIB_MAP_SPLICING
+  pg_temp_pending.merge(pg_temp_wanted);
+#else
+  pg_temp_pending.insert(make_move_iterator(begin(pg_temp_wanted)),
+			 make_move_iterator(end(pg_temp_wanted)));
+#endif
+  pg_temp_wanted.clear();
+}
+
+void ShardServices::requeue_pg_temp()
+{
+  unsigned old_wanted = pg_temp_wanted.size();
+  unsigned old_pending = pg_temp_pending.size();
+  _sent_pg_temp();
+  pg_temp_wanted.swap(pg_temp_pending);
+  logger().debug(
+    "{}: {} + {} -> {}",
+    __func__ ,
+    old_wanted,
+    old_pending,
+    pg_temp_wanted.size());
+}
+
+std::ostream& operator<<(
+  std::ostream& out,
+  const ShardServices::pg_temp_t& pg_temp)
+{
+  out << pg_temp.acting;
+  if (pg_temp.forced) {
+    out << " (forced)";
+  }
+  return out;
+}
+
+void ShardServices::send_pg_temp()
+{
+  if (pg_temp_wanted.empty())
+    return;
+  logger().debug("{}: {}", __func__, pg_temp_wanted);
+  boost::intrusive_ptr<MOSDPGTemp> ms[2] = {nullptr, nullptr};
+  for (auto& [pgid, pg_temp] : pg_temp_wanted) {
+    auto m = ms[pg_temp.forced];
+    if (!m) {
+      m = make_message< MOSDPGTemp >(osdmap->get_epoch());
```

Comment:
nit: do we need the extra whitechars here? In this file the convention is rather `make_message<MOSDPGInfo>`.